File: src/main/java/alfheim/client/core/proxy/ClientProxy.java
package alfheim.client.core.proxy;

import org.apache.commons.lang3.ArrayUtils;
import org.lwjgl.input.Keyboard;

import alexsocol.asjlib.ASJUtilities;
import alexsocol.asjlib.render.ASJShaderHelper;
import alfheim.AlfheimCore;
import alfheim.api.lib.*;
import alfheim.client.core.handler.CardinalSystemClient.TimeStopSystemClient;
import alfheim.client.core.handler.EventHandlerClient;
import alfheim.client.core.util.AlfheimBotaniaModifiersClient;
import alfheim.client.gui.*;
import alfheim.client.lib.LibResourceLocationsActual;
import alfheim.client.model.entity.*;
import alfheim.client.render.block.*;
import alfheim.client.render.entity.*;
import alfheim.client.render.item.*;
import alfheim.client.render.tile.*;
import alfheim.common.block.tile.*;
import alfheim.common.core.handler.EventHandler;
import alfheim.common.core.proxy.CommonProxy;
import alfheim.common.core.registry.AlfheimBlocks;
import alfheim.common.entity.*;
import alfheim.common.entity.boss.*;
import alfheim.common.entity.spell.*;
import alfheim.common.integration.travellersgear.handler.TGHandlerBotaniaRenderer;
import alfheim.common.lexicon.AlfheimLexiconData;
import cpw.mods.fml.client.registry.ClientRegistry;
import cpw.mods.fml.client.registry.RenderingRegistry;
import cpw.mods.fml.common.FMLCommonHandler;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.Gui;
import net.minecraft.client.settings.KeyBinding;
import net.minecraft.item.Item;
import net.minecraftforge.client.MinecraftForgeClient;
import net.minecraftforge.common.MinecraftForge;

public class ClientProxy extends CommonProxy {
	
	public static final KeyBinding keyCast = new KeyBinding("key.cast.desc", Keyboard.KEY_C, "key.categories.gameplay");
	public static final KeyBinding keyUnCast = new KeyBinding("key.uncast.desc", Keyboard.KEY_X, "key.categories.gameplay");
	public static final KeyBinding keyFlight = new KeyBinding("key.flight.desc", Keyboard.KEY_F, "key.categories.movement");
	public static final KeyBinding keySelMob = new KeyBinding("key.selmob.desc", Keyboard.KEY_R, "key.categories.gameplay");
	public static final KeyBinding keySelTeam = new KeyBinding("key.selteam.desc", Keyboard.KEY_T, "key.categories.gameplay");
	
	static {
		removeKeyBinding(keyCast);
		removeKeyBinding(keyFlight);
		removeKeyBinding(keySelMob);
		removeKeyBinding(keySelTeam);
	}
	
	private static final Gui guiIceLens = new GUIIceLens(Minecraft.getMinecraft());
	private static final Gui guiParty = new GUIParty(Minecraft.getMinecraft());
	private static final Gui guiRace = new GUIRace(Minecraft.getMinecraft());
	private static final Gui guiSpells = new GUISpells(Minecraft.getMinecraft());
	
	@Override
	public void preInit() {
		super.preInit();
		LibResourceLocationsActual.init();
	}
	
	@Override
	public void registerRenderThings() {
		LibRenderIDs.idAnyavil = RenderingRegistry.getNextAvailableRenderId();
		LibRenderIDs.idItemHolder = RenderingRegistry.getNextAvailableRenderId();
		LibRenderIDs.idPylon = RenderingRegistry.getNextAvailableRenderId();
		LibRenderIDs.idTransferer = RenderingRegistry.getNextAvailableRenderId();
		
		LibShaderIDs.idFire = ASJShaderHelper.createProgram(null, "shaders/fire.frag");
		LibShaderIDs.idGravity = ASJShaderHelper.createProgram(null, "shaders/gravity.frag");
		LibShaderIDs.idNoise = ASJShaderHelper.createProgram("shaders/position.vert", "shaders/noise4d.frag");
		LibShaderIDs.idShadow = ASJShaderHelper.createProgram(null, "shaders/shadow.frag");
		
	    MinecraftForgeClient.registerItemRenderer(Item.getItemFromBlock(AlfheimBlocks.anomaly), new RenderItemAnomaly());
		
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idAnyavil, new RenderBlockAnyavil());
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idItemHolder, new RenderBlockItemHolder());
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idPylon, new RenderBlockAlfheimPylons());
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idTransferer, new RenderBlockTransferer());
		
		ClientRegistry.bindTileEntitySpecialRenderer(TileAlfheimPortal.class, new RenderTileAlfheimPortal());
		ClientRegistry.bindTileEntitySpecialRenderer(TileAlfheimPylons.class, new RenderTileAlfheimPylons());
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnimatedTorch.class, new RenderTileAnimatedTorch());
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnomaly.class, new RenderTileAnomaly());
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnyavil.class, new RenderTileAnyavil());
		ClientRegistry.bindTileEntitySpecialRenderer(TileFlugelHead.class, new RenderTileFlugelHead());
		ClientRegistry.bindTileEntitySpecialRenderer(TileItemHolder.class, new RenderTileItemHolder());
		ClientRegistry.bindTileEntitySpecialRenderer(TileTradePortal.class, new RenderTileTradePortal());
		ClientRegistry.bindTileEntitySpecialRenderer(TileTransferer.class, new RenderTileTransferer());
		
		RenderingRegistry.registerEntityRenderingHandler(EntityAlfheimPixie.class, new RenderEntityAlfheimPixie());
		RenderingRegistry.registerEntityRenderingHandler(EntityElf.class, new RenderEntityElf(new ModelEntityElf(), 0.25F));
		RenderingRegistry.registerEntityRenderingHandler(EntityFlugel.class, new RenderEntityFlugel(new ModelEntityFlugel(), 0.25F));
		RenderingRegistry.registerEntityRenderingHandler(EntityRook.class, new RenderEntityRook(new ModelEntityRook(), 1.5F));
		RenderingRegistry.registerEntityRenderingHandler(EntityLightningMark.class, new RenderEntityLightningMark());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellHarp.class, new RenderEntityHarp());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellDriftingMine.class, new RenderEntityDriftingMine());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellGravityTrap.class, new RenderEntityGravityTrap());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellFenrirStorm.class, new RenderEntityFenrirStorm());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellMortar.class, new RenderEntityMortar());
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellWindBlade.class, new RenderEntityWindBlade());
	}
	
	@Override
	public void registerKeyBinds() {
		if (AlfheimCore.enableElvenStory) addESMKeyBinds();
		if (AlfheimCore.enableMMO) addMMOKeyBinds();
	}
	
	@Override
	public void initializeAndRegisterHandlers() {
		super.initializeAndRegisterHandlers();
		MinecraftForge.EVENT_BUS.register(new EventHandlerClient());
		FMLCommonHandler.instance().bus().register(new EventHandlerClient());
		if (AlfheimCore.TravellersGearLoaded) MinecraftForge.EVENT_BUS.register(new TGHandlerBotaniaRenderer());
		if (AlfheimCore.enableElvenStory) enableESMGUIs();
		if (AlfheimCore.enableMMO) enableMMOGUIs();
	}
	
	@Override
	public void postInit() {
		super.postInit();
		AlfheimBotaniaModifiersClient.postInit();
	}
	
	public static void toggelModes(boolean b, boolean esm, boolean mmo, boolean esmOld, boolean mmoOld) {
		if (b) toggleESM(esm, mmo, esmOld, mmoOld);
		else toggleMMO(esm, mmo, esmOld, mmoOld);
	}
	
	public static void enableESM() {
		if (AlfheimCore.enableElvenStory) return;
		AlfheimCore.enableElvenStory = true;
		AlfheimLexiconData.reEnableESM();
		enableESMGUIs();
		addESMKeyBinds();
		EventHandler.checkAddAttrs();
	}
	
	public static void disableESM() {
		if (!AlfheimCore.enableElvenStory) return;
		AlfheimCore.enableElvenStory = false;
		AlfheimLexiconData.disableESM();
		disableESMGUIs();
		removeESMKeyBinds();
		disableMMO();
	}
	
	public static void enableMMO() {
		if (AlfheimCore.enableMMO) return;
		AlfheimCore.enableMMO = true;
		AlfheimLexiconData.reEnableMMO();
		enableMMOGUIs();
		addMMOKeyBinds();
		enableESM();
	}
	
	public static void disableMMO() {
		if (!AlfheimCore.enableMMO) return;
		AlfheimCore.enableMMO = false;
		AlfheimLexiconData.disableMMO();
		disableMMOGUIs();
		removeMMOKeyBinds();
		TimeStopSystemClient.tsAreas.clear();
	}
	
	private static void toggleESM(boolean esm, boolean mmo, boolean esmOld, boolean mmoOld) {
		if (esmOld == esm) return;
		AlfheimCore.enableElvenStory = esm;
		
		if (esm) {
			AlfheimLexiconData.reEnableESM();
			addESMKeyBinds();
		} else {
			AlfheimLexiconData.disableESM();
			removeESMKeyBinds();
			if (mmoOld != mmo) toggleMMO(esm, mmo, esmOld, mmoOld);
		}
	}
	
	private static void toggleMMO(boolean esm, boolean mmo, boolean esmOld, boolean mmoOld) {
		if (mmoOld == mmo) return;
		AlfheimCore.enableMMO = mmo;
		
		if (mmo) {
			AlfheimLexiconData.reEnableMMO();
			enableMMOGUIs();
			addMMOKeyBinds();
			if (mmoOld != esm) toggleESM(esm, mmo, esmOld, mmoOld);
		} else {
			AlfheimLexiconData.disableMMO();
			disableMMOGUIs();
			removeMMOKeyBinds();
			TimeStopSystemClient.tsAreas.clear();
		}
	}
	
	private static void enableESMGUIs() {
		ASJUtilities.log("Registering ESM GUIs");
		MinecraftForge.EVENT_BUS.register(guiRace);
	}
	
	private static void disableESMGUIs() {
		ASJUtilities.log("Unregistering ESM GUIs");
		MinecraftForge.EVENT_BUS.unregister(guiRace);
	}
	
	private static void enableMMOGUIs() {
		ASJUtilities.log("Registering MMO GUIs");
		MinecraftForge.EVENT_BUS.register(guiIceLens);
		MinecraftForge.EVENT_BUS.register(guiParty);
		MinecraftForge.EVENT_BUS.register(guiSpells);
	}
	
	private static void disableMMOGUIs() {
		ASJUtilities.log("Unregistering MMO GUIs");
		MinecraftForge.EVENT_BUS.unregister(guiIceLens);
		MinecraftForge.EVENT_BUS.unregister(guiParty);
		MinecraftForge.EVENT_BUS.unregister(guiSpells);
	}
	
	private static void addESMKeyBinds() {
		addKeyBinding(keyFlight);
	}
	
	private static void removeESMKeyBinds() {
		unregisterKeyBinding(keyFlight);
	}
	
	private static void addMMOKeyBinds() {
		addKeyBinding(keyCast);
		addKeyBinding(keyUnCast);
		addKeyBinding(keySelMob);
		addKeyBinding(keySelTeam);
	}
	
	private static void removeMMOKeyBinds() {
		unregisterKeyBinding(keyCast);
		unregisterKeyBinding(keyUnCast);
		unregisterKeyBinding(keySelMob);
		unregisterKeyBinding(keySelTeam);
	}
	
	private static void unregisterKeyBinding(KeyBinding key) {
		removeKeyBinding(key);
		int id = ASJUtilities.indexOfComparableArray(Minecraft.getMinecraft().gameSettings.keyBindings, key);
		if (id < 0 || id > Minecraft.getMinecraft().gameSettings.keyBindings.length) return;
		Minecraft.getMinecraft().gameSettings.keyBindings = ArrayUtils.remove(Minecraft.getMinecraft().gameSettings.keyBindings, id);
	}
	
	private static void addKeyBinding(KeyBinding key) {
		key.setKeyCode(key.getKeyCodeDefault());
		ClientRegistry.registerKeyBinding(key);
	}
	
	private static void removeKeyBinding(KeyBinding key) {
		key.setKeyCode(0);
	}
}

File: src/main/java/alfheim/client/core/proxy/ClientProxy.kt
package alfheim.client.core.proxy

import alexsocol.asjlib.*
import alexsocol.asjlib.render.*
import alfheim.AlfheimCore
import alfheim.api.item.DoubleBoundItemRender
import alfheim.api.lib.*
import alfheim.client.core.handler.CardinalSystemClient.TimeStopSystemClient
import alfheim.client.core.handler.EventHandlerClient
import alfheim.client.core.util.AlfheimBotaniaModifiersClient
import alfheim.client.gui.*
import alfheim.client.lib.LibResourceLocationsActual
import alfheim.client.render.block.*
import alfheim.client.render.entity.*
import alfheim.client.render.item.*
import alfheim.client.render.particle.*
import alfheim.client.render.tile.*
import alfheim.common.block.AlfheimBlocks
import alfheim.common.block.tile.*
import alfheim.common.core.asm.hook.extender.FurnaceHandler
import alfheim.common.core.handler.*
import alfheim.common.core.proxy.CommonProxy
import alfheim.common.crafting.recipe.AlfheimRecipes
import alfheim.common.entity.*
import alfheim.common.entity.EntitySubspace
import alfheim.common.entity.EntitySubspaceSpear
import alfheim.common.entity.boss.*
import alfheim.common.entity.item.EntityItemImmortal
import alfheim.common.entity.spell.*
import alfheim.common.integration.thaumcraft.ThaumcraftAlfheimModule
import alfheim.common.integration.travellersgear.TGHandlerBotaniaRenderer
import alfheim.common.item.AlfheimItems
import alfheim.common.lexicon.AlfheimLexiconData
import cpw.mods.fml.client.registry.*
import net.minecraft.client.settings.KeyBinding
import net.minecraft.world.World
import net.minecraftforge.client.MinecraftForgeClient
import net.minecraftforge.common.MinecraftForge
import org.apache.commons.lang3.ArrayUtils
import org.lwjgl.input.Keyboard
import vazkii.botania.client.render.item.RenderLens
import vazkii.botania.common.Botania
import vazkii.botania.common.core.handler.ConfigHandler

class ClientProxy: CommonProxy() {
	
	override fun preInit() {
		super.preInit()
		LibResourceLocationsActual.init()
	}
	
	override fun registerRenderThings() {
		LibRenderIDs
		
		if (ConfigHandler.useShaders) {
			LibShaderIDs.idColor3d = ASJShaderHelper.createProgram("shaders/position.vert", "shaders/color3d.frag")
			LibShaderIDs.idGravity = ASJShaderHelper.createProgram(null, "shaders/gravity.frag")
			LibShaderIDs.idNoise = ASJShaderHelper.createProgram("shaders/position.vert", "shaders/noise4d.frag")
			LibShaderIDs.idShadow = ASJShaderHelper.createProgram(null, "shaders/shadow.frag")
		}
		
		ClientRegistry.registerKeyBinding(keyLolicorn)
		
		MinecraftForgeClient.registerItemRenderer(AlfheimItems.akashicRecords, RenderItemAkashicRecords)
		MinecraftForgeClient.registerItemRenderer(AlfheimBlocks.anomaly.toItem(), RenderItemAnomaly)
		MinecraftForgeClient.registerItemRenderer(AlfheimItems.fenrirClaws, RenderItemFenrirClaws)
		MinecraftForgeClient.registerItemRenderer(AlfheimItems.royalStaff, RenderItemRoyalStaff)
		
		if (!AlfheimConfigHandler.minimalGraphics)
			MinecraftForgeClient.registerItemRenderer(AlfheimItems.mjolnir, RenderItemMjolnir)
		
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idAnyavil, RenderBlockAnyavil)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idBarrel, RenderBlockBarrel)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idGrapeRedPlanted, RenderBlockGrapeRedPlanted)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idGrapeWhite, RenderBlockGrapeGreen)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idHarvester, RenderBlockAnomalyHarvester)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idManaAccelerator, RenderBlockItemHolder)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idPowerStone, RenderBlockPowerStone)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idPylon, RenderBlockAlfheimPylons)
		RenderingRegistry.registerBlockHandler(LibRenderIDs.idShrinePanel, RenderBlockShrinePanel)
		
		ClientRegistry.bindTileEntitySpecialRenderer(TileAlfheimPortal::class.java, RenderTileAlfheimPortal)
		ClientRegistry.bindTileEntitySpecialRenderer(TileAlfheimPylon::class.java, RenderTileAlfheimPylons)
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnimatedTorch::class.java, RenderTileAnimatedTorch)
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnomaly::class.java, RenderTileAnomaly)
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnomalyHarvester::class.java, RenderTileAnomalyHarvester)
		ClientRegistry.bindTileEntitySpecialRenderer(TileAnyavil::class.java, RenderTileAnyavil)
		ClientRegistry.bindTileEntitySpecialRenderer(TileBarrel::class.java, RenderTileBarrel)
		ClientRegistry.bindTileEntitySpecialRenderer(TileHeadFlugel::class.java, RenderTileHeadFlugel)
		ClientRegistry.bindTileEntitySpecialRenderer(TileHeadMiku::class.java, RenderTileHeadMiku)
		ClientRegistry.bindTileEntitySpecialRenderer(TileManaAccelerator::class.java, RenderTileManaAccelerator)
		ClientRegistry.bindTileEntitySpecialRenderer(TilePowerStone::class.java, RenderTilePowerStone)
		ClientRegistry.bindTileEntitySpecialRenderer(TileRaceSelector::class.java, RenderTileRaceSelector)
		ClientRegistry.bindTileEntitySpecialRenderer(TileTradePortal::class.java, RenderTileTradePortal)
		
		RenderingRegistry.registerEntityRenderingHandler(EntityAlfheimPixie::class.java, RenderEntityAlfheimPixie)
		RenderingRegistry.registerEntityRenderingHandler(EntityButterfly::class.java, RenderEntityButterfly)
		RenderingRegistry.registerEntityRenderingHandler(EntityElf::class.java, RenderEntityElf)
		RenderingRegistry.registerEntityRenderingHandler(EntityFenrir::class.java, RenderEntityFenrir)
		RenderingRegistry.registerEntityRenderingHandler(EntityFlugel::class.java, RenderEntityFlugel)
		RenderingRegistry.registerEntityRenderingHandler(EntityGleipnir::class.java, RenderEntityGleipnir)
		RenderingRegistry.registerEntityRenderingHandler(EntityItemImmortal::class.java, RenderEntityItemImmortal)
		RenderingRegistry.registerEntityRenderingHandler(EntityLightningMark::class.java, RenderEntityLightningMark)
		RenderingRegistry.registerEntityRenderingHandler(EntityLolicorn::class.java, RenderEntityLolicorn)
		RenderingRegistry.registerEntityRenderingHandler(EntityRook::class.java, RenderEntityRook)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellHarp::class.java, RenderEntityHarp)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellDriftingMine::class.java, RenderEntityDriftingMine)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellGravityTrap::class.java, RenderEntityGravityTrap)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellFenrirStorm::class.java, RenderEntityFenrirStorm)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellMortar::class.java, RenderEntityMortar)
		RenderingRegistry.registerEntityRenderingHandler(EntitySpellWindBlade::class.java, RenderEntityWindBlade)
		
		MinecraftForgeClient.registerItemRenderer(AlfheimItems.invisibleFlameLens, RenderLens())
		MinecraftForgeClient.registerItemRenderer(AlfheimItems.moonlightBow, RenderMoonBow)
		
		RenderingRegistry.registerBlockHandler(RenderBlockColoredDoubleGrass)
		RenderingRegistry.registerBlockHandler(MultipassRenderer)
		RenderingRegistry.registerBlockHandler(RenderBlockHopper)
		
		ClientRegistry.bindTileEntitySpecialRenderer(TileItemDisplay::class.java, RenderTileItemDisplay)
		ClientRegistry.bindTileEntitySpecialRenderer(TileEntityStar::class.java, RenderStar)
		
		RenderingRegistry.registerEntityRenderingHandler(EntityMjolnir::class.java, RenderEntityMjolnir())
		
		RenderingRegistry.registerEntityRenderingHandler(EntityThrownPotion::class.java, RenderEntityThrownPotion)
		RenderingRegistry.registerEntityRenderingHandler(EntityThrowableItem::class.java, RenderEntityThrownItem)
		
		RenderingRegistry.registerEntityRenderingHandler(EntitySubspace::class.java, RenderEntitySubspace)
		RenderingRegistry.registerEntityRenderingHandler(EntitySubspaceSpear::class.java, RenderEntitySubspaceSpear)
		
		RenderingRegistry.registerEntityRenderingHandler(EntityGrieferCreeper::class.java, RenderEntityGrieferCreeper)
		RenderingRegistry.registerEntityRenderingHandler(EntityVoidCreeper::class.java, RenderEntityGrieferCreeper)
	}
	
	override fun registerKeyBinds() {
		if (AlfheimConfigHandler.enableElvenStory) addESMKeyBinds()
		if (AlfheimConfigHandler.enableMMO) addMMOKeyBinds()
	}
	
	override fun initializeAndRegisterHandlers() {
		super.initializeAndRegisterHandlers()
		EventHandlerClient
		FurnaceHandler
		if (ConfigHandler.boundBlockWireframe) DoubleBoundItemRender
		if (AlfheimCore.TravellersGearLoaded) TGHandlerBotaniaRenderer
		if (AlfheimConfigHandler.enableElvenStory) enableESMGUIs()
		if (AlfheimConfigHandler.enableMMO) enableMMOGUIs()
		
		RenderPostShaders.allowShaders = !AlfheimConfigHandler.minimalGraphics
	}
	
	override fun postInit() {
		super.postInit()
		AlfheimBotaniaModifiersClient.postInit()
	}
	
	override fun featherFX(world: World, x: Double, y: Double, z: Double, color: Int, size: Float, lifetime: Float, distance: Float, must: Boolean) {
		if (mc.renderViewEntity != null && mc.effectRenderer != null) {
			val particle = EntityFeatherFx(world, x, y, z, color, size, lifetime)
			
			if (!must) {
				if (!doParticle(world)) return
				val distanceX: Double = mc.renderViewEntity.posX - particle.posX
				val distanceY: Double = mc.renderViewEntity.posY - particle.posY
				val distanceZ: Double = mc.renderViewEntity.posZ - particle.posZ
				if (distanceX * distanceX + distanceY * distanceY + distanceZ * distanceZ > distance * distance) {
					return
				}
			}
			
			mc.effectRenderer.addEffect(particle)
		}
	}
	
	override fun bloodFX(world: World, x: Double, y: Double, z: Double, lifetime: Int, size: Float, gravity: Float) {
		if (mc.renderViewEntity != null && mc.effectRenderer != null && doParticle(world)) {
			mc.effectRenderer.addEffect(EntityBloodFx(world, x, y, z, size, lifetime, gravity))
		}
	}
	
	fun doParticle(world: World): Boolean {
		return if (!world.isRemote) {
			false
		} else if (!ConfigHandler.useVanillaParticleLimiter) {
			true
		} else {
			var chance = 1f
			if (mc.gameSettings.particleSetting == 1) {
				chance = 0.6f
			} else if (mc.gameSettings.particleSetting == 2) {
				chance = 0.2f
			}
			chance == 1f || Math.random() < chance.D
		}
	}
	
	companion object {
		
		val keyLolicorn = KeyBinding("key.lolicorn.desc", Keyboard.KEY_L, "key.categories.alfheim")
		val keyESMAbility = KeyBinding("key.esmability.desc", Keyboard.KEY_M, "key.categories.alfheim")
		val keyFlight = KeyBinding("key.flight.desc", Keyboard.KEY_F, "key.categories.alfheim")
		val keyCast = KeyBinding("key.cast.desc", Keyboard.KEY_C, "key.categories.alfheim")
		val keyUnCast = KeyBinding("key.uncast.desc", Keyboard.KEY_X, "key.categories.alfheim")
		val keySelMob = KeyBinding("key.selmob.desc", Keyboard.KEY_R, "key.categories.alfheim")
		val keySelTeam = KeyBinding("key.selteam.desc", Keyboard.KEY_T, "key.categories.alfheim")
		
		init {
			removeKeyBinding(keyFlight)
			removeKeyBinding(keyESMAbility)
			removeKeyBinding(keyCast)
			removeKeyBinding(keyUnCast)
			removeKeyBinding(keySelMob)
			removeKeyBinding(keySelTeam)
		}
		
		private val guiIceLens = GUIIceLens()
		private val guiParty = GUIParty()
		private val guiRace = GUIRace()
		private val guiSpells = GUISpells()
		
		fun toggelModes(b: Boolean, esm: Boolean, mmo: Boolean, esmOld: Boolean, mmoOld: Boolean) {
			if (b)
				toggleESM(esm, mmo, esmOld, mmoOld)
			else
				toggleMMO(esm, mmo, esmOld, mmoOld)
		}
		
		fun enableESM() {
			if (AlfheimConfigHandler.enableElvenStory) return
			AlfheimConfigHandler.enableElvenStory = true
			AlfheimLexiconData.reEnableESM()
			if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.addESMRecipes()
			enableESMGUIs()
			addESMKeyBinds()
			ESMHandler.checkAddAttrs()
		}
		
		fun disableESM() {
			if (!AlfheimConfigHandler.enableElvenStory) return
			AlfheimConfigHandler.enableElvenStory = false
			AlfheimLexiconData.disableESM()
			if (Botania.thaumcraftLoaded) ThaumcraftAlfheimModule.removeESMRecipes()
			disableESMGUIs()
			removeESMKeyBinds()
			disableMMO()
		}
		
		fun enableMMO() {
			if (AlfheimConfigHandler.enableMMO) return
			AlfheimConfigHandler.enableMMO = true
			AlfheimLexiconData.reEnableMMO()
			AlfheimRecipes.addMMORecipes()
			enableMMOGUIs()
			addMMOKeyBinds()
			enableESM()
		}
		
		fun disableMMO() {
			if (!AlfheimConfigHandler.enableMMO) return
			AlfheimConfigHandler.enableMMO = false
			AlfheimLexiconData.disableMMO()
			AlfheimRecipes.removeMMORecipes()
			disableMMOGUIs()
			removeMMOKeyBinds()
			TimeStopSystemClient.clear()
		}
		
		private fun toggleESM(esm: Boolean, mmo: Boolean, esmOld: Boolean, mmoOld: Boolean) {
			if (esmOld == esm) return
			AlfheimConfigHandler.enableElvenStory = esm
			
			if (esm) {
				AlfheimLexiconData.reEnableESM()
				addESMKeyBinds()
			} else {
				AlfheimLexiconData.disableESM()
				removeESMKeyBinds()
				if (mmoOld != mmo) toggleMMO(false, mmo, esmOld, mmoOld)
			}
		}
		
		private fun toggleMMO(esm: Boolean, mmo: Boolean, esmOld: Boolean, mmoOld: Boolean) {
			if (mmoOld == mmo) return
			AlfheimConfigHandler.enableMMO = mmo
			
			if (mmo) {
				AlfheimLexiconData.reEnableMMO()
				enableMMOGUIs()
				addMMOKeyBinds()
				if (mmoOld != esm) toggleESM(esm, true, esmOld, mmoOld)
			} else {
				AlfheimLexiconData.disableMMO()
				disableMMOGUIs()
				removeMMOKeyBinds()
				TimeStopSystemClient.clear()
			}
		}
		
		private fun enableESMGUIs() {
			ASJUtilities.log("Registering ESM GUIs")
			MinecraftForge.EVENT_BUS.register(guiRace)
		}
		
		private fun disableESMGUIs() {
			ASJUtilities.log("Unregistering ESM GUIs")
			MinecraftForge.EVENT_BUS.unregister(guiRace)
		}
		
		private fun enableMMOGUIs() {
			ASJUtilities.log("Registering MMO GUIs")
			MinecraftForge.EVENT_BUS.register(guiIceLens)
			MinecraftForge.EVENT_BUS.register(guiParty)
			MinecraftForge.EVENT_BUS.register(guiSpells)
			
			MinecraftForge.EVENT_BUS.unregister(guiRace)
		}
		
		private fun disableMMOGUIs() {
			ASJUtilities.log("Unregistering MMO GUIs")
			MinecraftForge.EVENT_BUS.unregister(guiIceLens)
			MinecraftForge.EVENT_BUS.unregister(guiParty)
			MinecraftForge.EVENT_BUS.unregister(guiSpells)
			
			MinecraftForge.EVENT_BUS.register(guiRace)
		}
		
		private fun addESMKeyBinds() {
			addKeyBinding(keyFlight)
			addKeyBinding(keyESMAbility)
			
			KeyBinding.resetKeyBindingArrayAndHash()
		}
		
		private fun removeESMKeyBinds() {
			unregisterKeyBinding(keyFlight)
			unregisterKeyBinding(keyESMAbility)
			
			KeyBinding.resetKeyBindingArrayAndHash()
		}
		
		private fun addMMOKeyBinds() {
			addKeyBinding(keyCast)
			addKeyBinding(keyUnCast)
			addKeyBinding(keySelMob)
			addKeyBinding(keySelTeam)
			
			KeyBinding.resetKeyBindingArrayAndHash()
		}
		
		private fun removeMMOKeyBinds() {
			unregisterKeyBinding(keyCast)
			unregisterKeyBinding(keyUnCast)
			unregisterKeyBinding(keySelMob)
			unregisterKeyBinding(keySelTeam)
			
			KeyBinding.resetKeyBindingArrayAndHash()
		}
		
		private fun unregisterKeyBinding(key: KeyBinding) {
			removeKeyBinding(key)
			val id = ASJUtilities.indexOfComparableArray(mc.gameSettings.keyBindings, key)
			if (id < 0 || id > mc.gameSettings.keyBindings.size) return
			mc.gameSettings.keyBindings = ArrayUtils.remove(mc.gameSettings.keyBindings, id)
		}
		
		private fun addKeyBinding(key: KeyBinding) {
			key.keyCode = key.keyCodeDefault
			ClientRegistry.registerKeyBinding(key)
		}
		
		private fun removeKeyBinding(key: KeyBinding) {
			key.keyCode = 0
		}
	}
}